### DISCUSSION TITLE
help with adding new diagnostics

### DISCUSSION THREAD

**Discussion by alanjhewitt** [MetOffice/simulation-systems#278] [LFRic]
I am working on the lfric_apps side first to return a set value, later i will return actual values from UKCA code.

I am trying to work out how to include my new diagnostic field within the aerosol_fields field collection.
I get the following clean fail message from field_collection_mod.f90 (in the lfric repo).
ERROR: get_field: No 64-bit field [aod_ukca_ait_sol] in field collection: aerosol_fields.

I got the following error at revision r5169 . 
https://code.metoffice.gov.uk/trac/lfric_apps/log/main/branches/dev/alanjhewitt/r4747_AOD

I did think maybe i had to initialise these input files, so I just made a change at r5206 , but I dont think that would fix the issue I am having
https://code.metoffice.gov.uk/trac/lfric_apps/changeset/5206/main/branches/dev/alanjhewitt/r4747_AOD

help would be appricated

**Comment by mcdalvi** [comment]
Hi Alan,

If this is a prognostic (i.e. needs to be active between timesteps, populated or used by different sections) then it has to be added to aerosol_fields in create_physics__prognostics routine, see example of convection fields here
https://code.metoffice.gov.uk/trac/lfric_apps/changeset/4400/main/trunk/science/gungho/source/driver/create_physics_prognostics_mod.F90

If this is purely a diagnostic then there is a specific way for diagnostic processing in LFRic which happens at the algorithm level, but is not currently set up for UKCA:
https://code.metoffice.gov.uk/trac/lfric_apps/browser/main/trunk/science/um_physics_interface/source/diagnostics

**Comment by iboutle** [comment]
Hi Alan,

As Mohit says, we shouldn't be adding diagnostics to field collections, these have their own infrastructure as they only want memory allocating to them when the diagnostic is requested.

Essentially all you should need is:

mydiag_flag = init_diag(mydiag, 'mysec__mydiag')

where:
- mydiag_flag will give you a true/false flag of whether the diagnostics has been requested
- mydiag is the field to put your diagnostic data into (only fully allocated if the diagnostic is requested, minimal size otherwise)
- mysec_mydiag is the name in the xml file for the diagnostic

You would then output the diagnostic with:

if (mydiag_flag) call mydiag%write_field()

**Reply by iboutle** [reply]
I updated the diagnostic section here with some more details:

https://code.metoffice.gov.uk/trac/lfric/wiki/GhaspSupport/PhysicsHandsOn#a6.Outputtingnewdiagnostics
